File: solution/kotlin/week06-exercise2/src/main/kotlin/report/PdfReportGenerator.kt
package report

import arrow.core.Either
import arrow.core.right

class PdfReportGenerator : CanGenerateReport {
    override fun generateReport(data: List<ReportData>): Either<ReportError, String> {
        val report = buildString {
            appendLine("Starting PDF Report Generation...")
            appendLine("PDF Report Title: Comprehensive Data Report")
            appendLine("------------------------------------------------")
            data.forEach { d -> appendLine("Data ID: ${d.id} | Data Value: ${d.value} | Description: ${d.description}") }
            appendLine("------------------------------------------------")
            append("PDF Report Generated Successfully.")
        }
        return report.right()
    }
}

File: solution/c#/week06-legacy-code/ReportGenerator/PdfReportGenerator.cs
using System.Text;
using LanguageExt;

namespace ReportGenerator
{
    public class PdfReportGenerator : ICanGenerateReport
    {
        public Either<ReportError, string> GenerateReport(Seq<ReportData> data)
        {
            var sb = new StringBuilder();
            sb.AppendLine("Starting PDF Report Generation...");
            sb.AppendLine("PDF Report Title: Comprehensive Data Report");
            sb.AppendLine("------------------------------------------------");
            foreach (var d in data)
            {
                sb.AppendLine($"Data ID: {d.Id} | Data Value: {d.Value} | Description: {d.Description}");
            }

            sb.AppendLine("------------------------------------------------");
            sb.Append("PDF Report Generated Successfully.");
            return Prelude.Right<ReportError, string>(sb.ToString());
        }
    }
}